query/src/main/kotlin/com/linecorp/kotlinjdsl/query/CriteriaMutableQuerySpec.kt

```diff
@@ -0,0 +1,24 @@
+package com.linecorp.kotlinjdsl.query
+
+import com.linecorp.kotlinjdsl.query.clause.from.FromClause
+import com.linecorp.kotlinjdsl.query.clause.from.SimpleAssociatedJoinClause
+import com.linecorp.kotlinjdsl.query.clause.hint.JpaQueryHintClause
+import com.linecorp.kotlinjdsl.query.clause.hint.SqlQueryHintClause
+import com.linecorp.kotlinjdsl.query.clause.where.CriteriaQueryWhereClause
+
+interface CriteriaMutableQuerySpec<T> {
+    val targetEntity: Class<T>
+    val from: FromClause
+    val associate: SimpleAssociatedJoinClause
+    val where: CriteriaQueryWhereClause
+    val jpaHint: JpaQueryHintClause
+    val sqlHint: SqlQueryHintClause
+}
+
+/**
+ * According to the current specification, CriteriaMutableQuerySpec and CriteriaDeleteQuerySpec are the same.
+ * so instead of creating an interface, create a typealias.
+ * If the specifications change in the future, separate them into separate interfaces.
+ * This is because CriteriaUpdateQuerySpec inherits from CriteriaMutableQuerySpec.
+ */
+typealias CriteriaDeleteQuerySpec<T> = CriteriaMutableQuerySpec<T>
```

Comment:
-- english
It would be better to divide this from now on and combine it if there is anything to combine later.

Currently, if we look at the field of this Spec, it is a spec that can be used in Select Spec except for associate. Since it can be separated like WhereDsl in the future, I think it's better to separate it first now.

-- korean
이 부분은 지금부터 나누고 오히려 나중에 합칠 부분이 있으면 합치는 것이 좋을 것 같습니다.

현재 이 Spec의 필드를 보면 associate를 제외하고는 Select Spec에서도 쓰일 수 있는 스펙입니다. 추후 WhereDsl 과 같이 분리할 수도 있는 것이기 때문에 지금은 먼저 분리하는 것이 좋다고 판단합니다.

Reply:
well. In fact, I think that the current measures are not bad because I think that this part is a specification that has no room for further additions or changes in the first place. Is there any other reason why the specification of select, update, and delete should be changed after the structure? I think there will be very few. So I don't think this structure is bad. There are some parts that cannot be recycled due to the creation of a new interface.
Or, instead of typealias, it might be a way to create a delete that implements the Mutable interface as it is.

-- korean
글쎄요. 사실 이부분은 애초에 더 추가되거나 변경될 여지가 전혀 없는 스펙이라 생각해서 지금의 조치가 나쁘지 않다고 생각합니다. 지금 구조 이후에 select, update, delete 의 스펙이 변경이 되어야할 이유가 또 있을까요? 거의 없을거라 생각합니다. 그러므로 이 구조가 나쁘다고 생각이 들진 않습니다. 인터페이스를 새로 만듦으로 인해 재활용이 불가능한 부분도 생기기도 하고요
혹은 typealias 대신에 Mutable  인터페이스를 그대로 구현한 delete 를 만드는것도 방법이긴 하겠네요.

Reply:
And since typealias is already used, the change on the side of the user is not changed by creating a new interface. I think typealias is a sufficient keyword for that purpose.

-- korean
그리고 이미 typealias 를 쓰고 있기 때문에 사용하는 측에서의 변화는 인터페이스를 새로 만든다고해서 달라지는것도 없습니다. typealias 는 그런 용도로 사용하기에 충분한 키워드라고 생각합니다.

Reply:
Or, if the name is Mutable, which is a problem, I think it would be okay to use a general name such as Base. I'm curious about your thoughts.

-- korean
혹은 이름이 Mutable 이어서 문제가 된다면 Base 와 같은 일반적인 명칭으로 해도 괜찮을것 같은데요. 생각이 궁금합니다.

Reply:
-- english
It seems more so because there is currently no purpose of this interface.

For what purpose was the CriticalQuerySpec interface created?

-- korean
이 인터페이스의 목적성이 현재 없기 때문에 더 그런 것 같습니다.

CriteriaMutableQuerySpec 인터페이스는 어떠한 목적을 위해 만들어진 것인가요?

Reply:
It was created for deduplication to solve the problem of needing to know each and every detailed interface inheritance. Regarding the overlapping part in select , if select inherits mutable (base if name is changed) interface, I think that the additional part of each dml implementation interface will be more clearly seen.

-- korean
각 세부 인터페이스 상속을 다 알아야 하는 문제를 해결하기 위해 중복 제거용으로 만들었습니다. select 에서도 겹친다는 부분에 대해서는 select 도 mutable(이름을 바꾸게 되면 base) 인터페이스를 상속하도록 하면 각 dml 구현 인터페이스의 추가적인 부분이 뭔지 더 명확하게 보일 것이라 생각이 듭니다.

Reply:
-- english
I think the problem is that an interface only for the extraction of common parts.

Dsl interface, for example, structure does not have a base. If we made BaseDsl, it would be much harder to use than now.

As such, we don't know how this interface will change in the future. There is a common method, so it's just grouped together.

If the role of this interface has not been defined like this, some people will wonder if it is okay to use it elsewhere.

Like the Common AbstractCriteria we're discussing below.

-- korean
공통 부분의 추출만을 위한 인터페이스가 되는 것이 문제라고 생각합니다.

Dsl 인터페이스를 예를 들면 지금 Dsl 인터페이스 구조는 Base가 없습니다. 만약 BaseDsl을 만들었다면 지금보다 훨씬 사용하기 힘들 것입니다. 

이처럼 이 인터페이스가 앞으로 어떻게 변경될지 모릅니다. 공통 메소드가 존재하니 하나로 묶었을 뿐입니다.

이처럼 이 인터페이스가 가진 역할이 정의 되지 않았다면 다른 곳에 사용할 때에도 정말 사용해도 괜찮은지 의문을 가지게 될 것입니다. 

마치 아래에서 의논하고 있는 CommonAbstractCriteria 처럼요

Reply:
A very suitable example of how we combine interfaces of dsl and remove duplicates is something like SpringDataCriteriaQueryDsl .

-- korean
우리가 dsl 의 인터페이스를 조합하고 중복을 제거하는 예로 아주 적절한게 SpringDataCriteriaQueryDsl 와 같은게 있습니다.

Reply:
-- english
Interfaces with no role set like this are too likely to be a problem when used in lower layers.

-- korean
이처럼 역할이 정해지지 않은 인터페이스는 하위 레이어에서 사용할 때 문제가 될 가능성이 너무 많습니다.

Reply:
I don't think it's appropriate at all to say that the role isn't set.
Because I think it's a good practice to create a base to de-duplicate so that each sub-implementation knows minimal extra information.

The part of page, order by, etc. on the SpringData side is a problem with spring itself, and the actual mutable interface is a minimum agreement that cannot be further split there.

-- korean
역할이 정해지지 않았다는 말은 전혀 적합하다고 생각하진 않습니다.
중복을 제거하기 위한 base 를 만들어서 각 하위 구현체에서 최소한의 추가 정보만 알 수 있게 하는게 좋은 방식이라고 생각하기 때문입니다.

SpringData 쪽의 page, order by 등의 부분은 spring 자체에 대한 문제이고 실제 mutable 인터페이스는 저기서 더 쪼개질 수 없는 최소한의 합의점을 만들어놓은것입니다.

Reply:
-- english
There are too few cases that can be checked at this time to combine Update, Delete, and Select into one.

If an incorrect interface is defined this time, it may continue to affect subsequent compatibility.

Therefore, after discovering several cases in the future, we decided to define which interface to provide to the user and then define the parent interface.

-- korean
이번에 Update 및 Delete, Select를 하나로 묶기에는 현재 확인할 수 있는 케이스가 너무 적습니다.

이번에 잘못 인터페이스가 정의될 경우 이후의 하위 호환성에 계속 영향을 끼칠 수 있습니다.

그렇기 때문에 추후에 여러 케이스들을 발견한 후 어떤 인터페이스를 사용자에게 제공할지에 대해서 정의 후에 상위 인터페이스를 정의하기로 결정했습니다.